File: Pie02/Pie02IronPython/Rules/PieGrammar.py
from Irony import *
from Irony.Parsing import *

from ClassRules import ClassRules
from ImportRules import ImportRules
from NamespaceRules import NamespaceRules
from MethodDeclarationRules import *
from OperatorRules import *
from LiteralRules import *

class PieGrammar(Grammar):
    """description of class"""

    def __init__(self):

        self.Class = KeyTerm("class", "class")
        self.Colon = KeyTerm(":", ":")
        self.Comma = KeyTerm(",", ",")
        self.Dot = KeyTerm(".", ".")
        self.Import = KeyTerm("import", "import")
        self.Interface = KeyTerm("interface", "interface")
        self.Internal = KeyTerm("internal", "internal")
        self.Final = KeyTerm("final", "final")
        self.Namespace = KeyTerm("namespace", "namespace")
        self.Module = KeyTerm("module", "module")
        self.Private = KeyTerm("private", "private")
        self.Public = KeyTerm("public", "public")
        self.Shared = KeyTerm("shared", "shared")
        self.Struct = KeyTerm("struct", "struct")
        self.Void = KeyTerm("void", "void")

        self.LParens = KeyTerm("(", "(")
        self.GThan = KeyTerm(">", ">")
        self.LThan = KeyTerm("<", "<")
        self.RParens = KeyTerm(")", ")")

        self.Identifier = IdentifierTerminal("identifier")
        self.IdentifierList = NonTerminal("identifier_list")
        self.IdentifierList.Rule = self.MakeStarRule(self.IdentifierList, self.Comma, self.Identifier)
      
        self.QualifiedIdentifier = NonTerminal("qualified_identifier")
        self.QualifiedIdentifier.Rule = self.MakePlusRule(self.QualifiedIdentifier, self.Dot, self.Identifier)

        self.Modifier = NonTerminal("modifier")
        self.Modifier.Rule = self.Internal | self.Private | self.Public | self.Shared | self.Final
        self.ModifierList = NonTerminal("modifier_list")
        self.ModifierList.Rule = self.MakeStarRule(self.ModifierList, self.Modifier)

        self.GenericList = NonTerminal("generic_list")
        self.GenericList.Rule = self.LThan + self.IdentifierList + self.GThan
        self.GenericListOpt = NonTerminal("generic_list_opt")
        self.GenericListOpt.Rule = self.GenericList | self.Empty

        self.LiteralRules = LiteralRules(self)

        self.ImportRules = ImportRules(self)

        self.Expression = NonTerminal("expression")
        self.Expression.Rule = self.QualifiedIdentifier
        self.Expression.Rule |= self.LiteralRules.Literal

        self.OperatorRules = OperatorRules(self)
        self.StatementExpression = NonTerminal("statement_expression")
        self.StatementExpression.Rule = self.OperatorRules.Assignment + self.Eos

        self.MethodDeclarationRules = MethodDeclarationRules(self)

        self.ClassRules = ClassRules(self)

        self.NamespaceRules = NamespaceRules(self)



        self.CompileUnit = NonTerminal("compile_unit")
        self.CompileUnit.Rule = self.NamespaceRules.NamespaceMemberList



        self.Root = self.CompileUnit

    
    def CreateTokenFilters(self, language, filters):
        outlineFilter = CodeOutlineFilter(language.GrammarData, OutlineOptions.ProduceIndents | OutlineOptions.CheckOperator | OutlineOptions.CheckBraces, self.ToTerm("_"))
        filters.Add(outlineFilter)

File: Pie02/Pie02/Parser/PieGrammar.vb
Imports Irony.Parsing

Public Class PieGrammar
    Inherits Grammar

    Public ClassT As New KeyTerm("class", "class")
    Public Colon As New KeyTerm(":", ":")
    Public Comma As New KeyTerm(",", ",")
    Public Dot As New KeyTerm(".", ".")
    Public Identifier As New IdentifierTerminal("identifier")
    Public Import As New KeyTerm("import", "import")
    Public NamespaceT As New KeyTerm("namespace", "namespace")

    Public ImportRules As ImportRules
    Public NamespaceRules As NamespaceRules
    Public ClassRules As ClassRules

    Public QualifiedIdentifier As New NonTerminal("qualifiedIdentifier")

    Public Sub New()

        QualifiedIdentifier.Rule = MakePlusRule(QualifiedIdentifier, Dot, Identifier)

        ImportRules = New ImportRules(Me)
        ClassRules = New ClassRules(Me)
        NamespaceRules = New NamespaceRules(Me)

        Dim compileUnit = New NonTerminal("compileUnit")
        compileUnit.Rule = NamespaceRules.NamespaceMemberList

        Me.Root = compileUnit
    End Sub

    Public Overrides Sub CreateTokenFilters(language As LanguageData, filters As TokenFilterList)
        Dim outlineFilter = New CodeOutlineFilter(language.GrammarData,
              OutlineOptions.ProduceIndents Or OutlineOptions.CheckOperator Or OutlineOptions.CheckBraces, ToTerm("\"))
        filters.Add(outlineFilter)
    End Sub


End Class
